com.unity.ml-agents.extensions/Runtime/Match3/Match3Sensor.cs

```diff
@@ -245,38 +254,86 @@ static void DestroyTexture(Texture2D texture)
     internal class OneHotToTextureUtil
     {
         Color[] m_Colors;
-        int m_Height;
-        int m_Width;
+        int m_MaxHeight;
+        int m_MaxWidth;
         private static Color[] s_OneHotColors = { Color.red, Color.green, Color.blue };
 
-        public OneHotToTextureUtil(int height, int width)
+        public OneHotToTextureUtil(int maxHeight, int maxWidth)
         {
-            m_Colors = new Color[height * width];
-            m_Height = height;
-            m_Width = width;
+            m_Colors = new Color[maxHeight * maxWidth];
+            m_MaxHeight = maxHeight;
+            m_MaxWidth = maxWidth;
         }
 
-        public void EncodeToTexture(GridValueProvider gridValueProvider, Texture2D texture, int channelOffset)
+        public void EncodeToTexture(
+            GridValueProvider gridValueProvider,
+            Texture2D texture,
+            int channelOffset,
+            int currentHeight,
+            int currentWidth
+            )
         {
             var i = 0;
             // There's an implicit flip converting to PNG from texture, so make sure we
             // counteract that when forming the texture by iterating through h in reverse.
-            for (var h = m_Height - 1; h >= 0; h--)
+            for (var h = m_MaxHeight - 1; h >= 0; h--)
             {
-                for (var w = 0; w < m_Width; w++)
+                for (var w = 0; w < m_MaxWidth; w++)
                 {
-                    int oneHotValue = gridValueProvider(h, w);
-                    if (oneHotValue < channelOffset || oneHotValue >= channelOffset + 3)
-                    {
-                        m_Colors[i++] = Color.black;
-                    }
-                    else
+                    var colorVal = Color.black;
+                    if (h < currentHeight && w < currentWidth)
```

Comment:
again, maybe we can get rid of this branch by multiplying out with a combo of min/max